File: packages/batch/src/schema.ts
import * as v from "valibot";
import type { ToolDefinition } from "@agent-sh/harness-core";
import type {
  BatchParams,
  BatchTargetSubdirs,
  BatchTargetGlob,
  BatchTargetExplicit,
} from "./types.js";

// Target schemas

export const BatchTargetSubdirsSchema = v.strictObject({
  kind: v.literal("subdirs"),
  path: v.pipe(v.string(), v.minLength(1, "path is required")),
  name_filter: v.optional(v.string()),
});

export const BatchTargetGlobSchema = v.strictObject({
  kind: v.literal("glob"),
  pattern: v.pipe(v.string(), v.minLength(1, "pattern is required")),
});

export const BatchTargetExplicitSchema = v.strictObject({
  kind: v.literal("explicit"),
  paths: v.pipe(
    v.array(v.pipe(v.string(), v.minLength(1, "path must not be empty"))),
    v.minLength(1, "paths must contain at least one entry"),
  ),
});

export const BatchTargetSchema = v.union([
  BatchTargetSubdirsSchema,
  BatchTargetGlobSchema,
  BatchTargetExplicitSchema,
]);

export type ParsedBatchTarget =
  | v.InferOutput<typeof BatchTargetSubdirsSchema>
  | v.InferOutput<typeof BatchTargetGlobSchema>
  | v.InferOutput<typeof BatchTargetExplicitSchema>;

export const BatchParamsSchema = v.strictObject({
  command: v.pipe(
    v.string(),
    v.minLength(1, "command is required"),
    v.maxLength(16384, "command exceeds 16384 bytes"),
  ),
  targets: BatchTargetSchema,
  mode: v.optional(v.picklist(["sequential", "parallel"])),
  max_concurrent: v.optional(
    v.pipe(v.number(), v.integer(), v.minValue(1, "max_concurrent must be >= 1")),
  ),
  timeout_secs: v.optional(
    v.pipe(
      v.number(),
      v.integer(),
      v.minValue(1, "timeout_secs must be >= 1"),
      v.maxValue(3600, "timeout_secs must be <= 3600 (1 hour)"),
    ),
  ),
  fail_fast: v.optional(v.boolean()),
  summary_only: v.optional(v.boolean()),
});

export type ParsedBatchParams = v.InferOutput<typeof BatchParamsSchema>;

/**
 * Alias table for common model mistakes.
 */
const KNOWN_PARAM_ALIASES: Record<string, string> = {
  repos:
    "unknown parameter 'repos'. Use 'targets' instead (e.g. targets: { kind: 'subdirs', path: '/home/avifenesh/projects' }).",
  directories:
    "unknown parameter 'directories'. Use 'targets' instead.",
  dirs: "unknown parameter 'dirs'. Use 'targets' instead.",
  paths:
    "unknown parameter 'paths'. Use 'targets: { kind: 'explicit', paths: [...] }' instead.",
  git_cmd: "unknown parameter 'git_cmd'. Use 'command' instead.",
  git_command: "unknown parameter 'git_command'. Use 'command' instead.",
  cmd: "unknown parameter 'cmd'. Use 'command' instead.",
  shell_cmd: "unknown parameter 'shell_cmd'. Use 'command' instead.",
  parallel:
    "unknown parameter 'parallel'. Use 'mode: 'parallel'' instead.",
  concurrent:
    "unknown parameter 'concurrent'. Use 'max_concurrent' instead.",
  timeout:
    "unknown parameter 'timeout'. Use 'timeout_secs' instead (seconds, not milliseconds).",
};

function checkAliases(input: unknown): string[] {
  if (input === null || typeof input !== "object") return [];
  const hints: string[] = [];
  for (const key of Object.keys(input as Record<string, unknown>)) {
    const hint = KNOWN_PARAM_ALIASES[key];
    if (hint) hints.push(hint);
  }
  return hints;
}

function makeAliasIssues(messages: string[]): v.BaseIssue<unknown>[] {
  return messages.map(
    (m) =>
      ({
        kind: "validation" as const,
        type: "custom" as const,
        input: undefined,
        expected: null,
        received: "unknown",
        message: m,
      }) as unknown as v.BaseIssue<unknown>,
  );
}

export function safeParseBatchParams(input: unknown):
  | { ok: true; value: BatchParams }
  | { ok: false; issues: v.BaseIssue<unknown>[] } {
  const aliases = checkAliases(input);
  if (aliases.length > 0) {
    return { ok: false, issues: makeAliasIssues(aliases) };
  }
  const result = v.safeParse(BatchParamsSchema, input);
  if (result.success) return { ok: true, value: result.output };
  return { ok: false, issues: result.issues };
}

// Tool definition exposed to the LLM.

export const BATCH_TOOL_NAME = "batch";

export const BATCH_TOOL_DESCRIPTION = `Execute a shell command across multiple directories. Use for batch git operations, bulk file operations, workspace builds, etc.

Supports three target modes:
- subdirs: run in each immediate subdirectory of a path
- glob: run in each path matching a glob pattern
- explicit: run in each explicitly listed path

Execution modes:
- sequential: run one at a time, accumulate results
- parallel: run concurrently (max_concurrent controls concurrency)

Use $TARGET in the command to reference the current directory.
Use fail_fast to stop on first failure.
Use summary_only to get counts instead of per-target output.`;

export const batchToolDefinition: ToolDefinition = {
  name: BATCH_TOOL_NAME,
  description: BATCH_TOOL_DESCRIPTION,
  inputSchema: {
    type: "object",
    properties: {
      command: {
        type: "string",
        description:
          "Shell command to run in each target. Use $TARGET to reference the current directory.",
      },
      targets: {
        oneOf: [
          {
            type: "object",
            properties: {
              kind: { const: "subdirs" },
              path: {
                type: "string",
                description: "Root directory to scan for subdirectories.",
              },
              name_filter: {
                type: "string",
                description: "Optional name filter (glob pattern for subdirectory names).",
              },
            },
            required: ["kind", "path"],
          },
          {
            type: "object",
            properties: {
              kind: { const: "glob" },
              pattern: {
                type: "string",
                description: "Glob pattern for target directories.",
              },
            },
            required: ["kind", "pattern"],
          },
          {
            type: "object",
            properties: {
              kind: { const: "explicit" },
              paths: {
                type: "array",
                items: { type: "string" },
                description: "Explicit list of directory paths.",
              },
            },
            required: ["kind", "paths"],
          },
        ],
      },
      mode: {
        type: "string",
        enum: ["sequential", "parallel"],
        description: "Execution mode. Default: sequential.",
      },
      max_concurrent: {
        type: "integer",
        minimum: 1,
        description: "Max concurrent commands for parallel mode (default 4).",
      },
      timeout_secs: {
        type: "integer",
        minimum: 1,
        maximum: 3600,
        description: "Timeout per command in seconds (default 120).",
      },
      fail_fast: {
        type: "boolean",
        description: "Stop on first failure (default false).",
      },
      summary_only: {
        type: "boolean",
        description: "Return only summary counts, not per-target output (default false).",
      },
    },
    required: ["command", "targets"],
    additionalProperties: false,
  },
};

File: crates/batch/src/schema.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

/// Targets for batch operations. Uses `kind`-based tagging to match
/// the Node schema shape: `{ kind: "subdirs", path: "..." }`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", tag = "kind")]
pub enum BatchTarget {
    /// Run the command in each immediate subdirectory of the given path.
    Subdirs {
        /// Root directory to scan for subdirectories.
        path: String,
        /// Optional glob filter for subdirectory names (e.g., "*.git").
        #[serde(default, skip_serializing_if = "Option::is_none")]
        name_filter: Option<String>,
    },
    /// Run the command in each path matching the glob.
    Glob {
        /// Glob pattern (e.g., "~/projects/**/*.git").
        pattern: String,
    },
    /// Run the command in each explicitly listed path.
    Explicit {
        paths: Vec<String>,
    },
}

/// How to execute the batch command.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub enum BatchMode {
    /// Run sequentially, accumulating results.
    #[default]
    Sequential,
    /// Run in parallel (max_concurrent controls concurrency).
    Parallel,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BatchParams {
    /// Shell command to run in each target directory. Use `$TARGET` to
    /// reference the current directory path.
    pub command: String,
    /// What directories/paths to operate on.
    pub targets: BatchTarget,
    /// Execution mode.
    #[serde(default)]
    pub mode: BatchMode,
    /// Maximum concurrent commands for parallel mode (default 4).
    #[serde(default = "default_max_concurrent", alias = "max_concurrent")]
    pub max_concurrent: usize,
    /// Timeout per command in seconds (default 120).
    #[serde(default = "default_timeout_secs", alias = "timeout_secs")]
    pub timeout_secs: u64,
    /// If true, stop on first failure.
    #[serde(default, alias = "fail_fast")]
    pub fail_fast: bool,
    /// If true, return only a summary (not full per-target output).
    #[serde(default, alias = "summary_only")]
    pub summary_only: bool,
}

fn default_max_concurrent() -> usize {
    4
}

fn default_timeout_secs() -> u64 {
    120
}

#[derive(Debug, Clone, thiserror::Error)]
pub enum BatchParseError {
    #[error("{0}")]
    Message(String),
}

pub fn safe_parse_batch_params(input: &Value) -> Result<BatchParams, BatchParseError> {
    let parsed: BatchParams = serde_json::from_value(input.clone())
        .map_err(|e| BatchParseError::Message(e.to_string()))?;
    if parsed.command.is_empty() {
        return Err(BatchParseError::Message(
            "command must not be empty".to_string(),
        ));
    }
    Ok(parsed)
}
